### PR TITLE
fix(ouroboros): stop caller deadline racing its own attempt timer

### LINKED ISSUE #4154: test(ouroboros): endorser-block deadline test races two timeouts
`ouroboros.TestFetchEndorserBlockByPointDeadlineDoesNotCoolDownPeer` fails intermittently on CI, most often on Windows, across pull requests whose diffs do not touch `ouroboros/`.

## Failure

```
--- FAIL: TestFetchEndorserBlockByPointDeadlineDoesNotCoolDownPeer (2.00s)
    leios_certified_recovery_test.go:654:
        Error: Should be zero, but was 1
```

Line 654 is `require.Zero(t, guard.consecutiveFailures.Load())` — the assertion that an expired *caller* deadline must not be recorded as a *peer* failure, which is the behaviour the test exists to protect (see #2819).

## Observed

| Pull request | Job | Diff scope |
| --- | --- | --- |
| #3853 | `go-test (macOS)`, then `go-test (Windows)` on a re-run | `internal/koiosparity/`, `ARCHITECTURE.md` |
| #3929 | `go-test (Windows)` | `internal/config/`, examples, docs |

Neither branch touches `ouroboros/`. The duration is `2.00s` every time.

It did **not** reproduce locally: `go test ./ouroboros/ -cpu=1 -count=1` on an unloaded arm64 macOS host, 6 runs at clean `origin/main` (`2d6441ae`), 0 failures. That negative result is consistent with the failure needing a slow or loaded runner rather than a particular interleaving — `-cpu=1` reproduces the other tests in this class near-deterministically and does nothing here.

## Likely mechanism, not yet confirmed

The test relies on an ordering between two timeouts. It waits up to 2s for the fetch to reach its stalled state, then expects the caller's context deadline to expire and `FetchEndorserBlockByPoint` to return `context.DeadlineExceeded` **before** anything peer-level times out, so the guard's failure counter stays at zero.

On a runner slow enough, that ordering can invert: a peer-level timeout fires first, the failure is recorded legitimately, and the assertion sees `1`. If so the production behaviour is correct and the test is asserting a race between two real-time budgets.

This is a hypothesis from reading `leios_certified_recovery_test.go:630-656`, not something I have demonstrated. Confirming it means establishing which timeout fires first in a failing run — worth doing before changing anything, because the alternative reading is that a caller-deadline path really does increment the counter on some code path, which would be a production defect rather than a test one.

## Suggested direction

If the ordering inversion is confirmed: make the test assert the property without racing two wall-clock budgets — drive the deadline from an injectable clock, or make the peer-level timeout unreachable within the test rather than merely slower than the caller's.

Please do not fix it by widening the 2s window, by dropping `t.Parallel()`, or by skipping on Windows: the first re-creates the same race with a bigger constant, and the second gives back the CI time #4131 was added to reclaim.

## Related

Same class as #4146 (`ledger` scheduler and slot-clock tests asserting on hard real-time windows) and #2265 (`time.Sleep` synchronisation). Distinct from #4149, which is shared global state rather than timing, and from #4145, which was a missing happens-before edge.

`ouroboros/leios_certified_recovery_test.go` was last modified by `2bcc8c4a` (#4131, "perf(test): run independent tests in parallel"), the same commit implicated in all three of the above.

### PR DESCRIPTION
**Pull request by chrisguiney**
`FetchEndorserBlockByPoint` computes the last (or only) backfill candidate's
per-attempt deadline as its own `time.Now().Add(budget)` — a second,
independent clock read separate from the caller's `ctx.Deadline()`. Go's
`context.WithDeadline` only reuses a parent's own cancellation timer when the
parent's deadline is *strictly* earlier than the requested one
(`cur.Before(d)`); an equal (or, under scheduling skew, marginally earlier)
derived deadline instead arms a second, independent timer alongside the
caller's. When that second timer fires first, `fetchEndorserBlockOnConn`'s
cooldown classifier reads the *caller's* context, sees `ctx.Err() == nil`,
and records the failure against the peer even though the caller's own
deadline is what ended the attempt.

Operational impact: a caller with a tight budget can cool down a healthy
Leios-fetch peer under load, degrading fetch throughput.

Fix, in `ouroboros/leios_backfill.go`:

- `FetchEndorserBlockByPoint` derives the last candidate's `attemptDeadline`
  from the same `time.Now()` read used for its remaining-budget calculation,
  so that deadline is exactly `overall` by construction instead of
  approximated by a second clock read.
- `leiosFetchRequestContext` gets an explicit boundary check — extracted as
  `leiosFetchRequestContextReusesParent` — that derives the request context
  via `context.WithCancel(parent)` whenever the parent's deadline is not
  later than the requested one, covering the equal-deadline case that
  `context.WithDeadline`'s own strict-inequality shortcut misses.
- A genuinely earlier per-attempt deadline (a truncated multi-candidate
  attempt) still gets its own timer, so failover truncation (#2819, #3552)
  is unchanged.

Confirmed causally: a temporary clock-skew injection on the original
`time.Now().Add(budget)` reproduced `guard.consecutiveFailures == 1` with
`err` still `errors.Is(context.DeadlineExceeded)` — the exact CI failure
shape — and was reverted before this fix was implemented.

competing hypotheses for this flake; this change resolves it in favor of
production defect.

---
## Summary by cubic
Fixes the Leios backfill deadline handling so a caller's own timeout can no longer be misattributed to a healthy peer and trigger an erroneous cooldown. The last (or only) backfill candidate's per-attempt deadline is now derived from the same clock read as the remaining budget, and the request context reuses the caller's cancellation when the deadlines are equal.

**Bug Fixes**
- An equal (or, under skew, marginally off) per-attempt deadline no longer arms a second independent timer; `leiosFetchRequestContext` now returns `context.WithCancel(parent)` in that case.
- A genuinely earlier per-attempt deadline still gets its own timer, preserving multi-candidate failover truncation.
- Adds a test pinning the equal-deadline boundary that previously made the deadline-cooldown check flaky on Windows CI.

<sup>Written for commit 9014b591872342905b5b93d55248e740e99328d3. Summary will update on new commits.</sup>

Closes #4154

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved timeout handling for Leios backfill requests by consistently calculating per-attempt deadlines.
  * Prevented unnecessary independent timers when an existing parent deadline is sooner or equal, improving cancellation behavior.

* **Tests**
  * Added coverage for deadline boundary cases and verified that redundant timers are not created.

### CODE REVIEW

**Review by cryptodj413 — DISMISSED** [review]

File: ouroboros/leios_certified_recovery_test.go [hunk 1]
```diff
@@ -655,6 +655,85 @@ func TestFetchEndorserBlockByPointDeadlineDoesNotCoolDownPeer(t *testing.T) {
...
+// fetchEndorserBlockOnConn's ctx.Err() check (which reads the caller's
+// context specifically) observes it as not-yet-expired and misattributes the
+// caller's own deadline to the peer, incrementing consecutiveFailures on a
+// connection that did nothing wrong.
+//
+// This does not race real timers to prove the point: whether an independent
+// timer got armed is not black-box observable without either racing it or
+// reflecting into context's unexported types, so the decision that prevents
+// it is pinned directly instead.
+func TestLeiosFetchRequestContextReusesParentAtEqualDeadline(t *testing.T) {
```
Comment (wolf31o2) [inline ouroboros/leios_certified_recovery_test.go:679]: This regression test only exercises `leiosFetchRequestContextReusesParent`; it never calls the production `leiosFetchRequestContext` used by `leios_backfill.go`. Replacing that production call with unconditional `context.WithDeadline` would reintroduce the caller-deadline timer race while this test still passes. Add a behavioral test through the production helper proving equal or earlier parent deadlines do not arm an independent timer.

**Review by wolf31o2 — CHANGES_REQUESTED** [review]
The implementation is otherwise reasonable, but this test/reliability gap should be fixed.

File: ouroboros/leios_certified_recovery_test.go (hunk 1 above)
Comment (chrisguiney) [reply ouroboros/leios_certified_recovery_test.go:679]: Added `TestLeiosFetchRequestContextDoesNotArmIndependentTimerAtEqualOrEarlierParentDeadline`, which calls `leiosFetchRequestContext` directly under `testing/synctest`. The parent context reports a deadline through a `Deadline()`-overriding wrapper but carries no timer of its own, so any expiry the returned request context exhibits at the requested deadline can only come from a timer `leiosFetchRequestContext` itself armed.

Reverted `leiosFetchRequestContext` to unconditional `context.WithDeadline` in place: the equal-deadline subtest fails 10/10 runs under `-race` with `context deadline exceeded`. Reverting the mutation restores a pass.

File: ouroboros/leios_certified_recovery_test.go [hunk 2]
```diff
@@ -655,6 +656,204 @@ func TestFetchEndorserBlockByPointDeadlineDoesNotCoolDownPeer(t *testing.T) {
...
+// (nothing else in the bubble is running to race against).
+//
+// Early parent cancellation does not discriminate the two branches (an
+// independent timerCtx derived from the parent is cancelled by an early
+// parent cancel too, since it is still registered as the parent's child) so
+// this test does not rely on that; it relies on the parent surviving past the
+// requested deadline uncancelled instead.
+func TestLeiosFetchRequestContextDoesNotArmIndependentTimerAtEqualOrEarlierParentDeadline(
+	t *testing.T,
+) {
```
[the remark this answers is not in the document]
Comment (chrisguiney) [reply ouroboros/leios_certified_recovery_test.go:797]: Fixed in 4e5b61e9: added `t.Parallel()` as the first statement.

File: ouroboros/leios_certified_recovery_test.go [hunk 3]
```diff
@@ -655,6 +656,204 @@ func TestFetchEndorserBlockByPointDeadlineDoesNotCoolDownPeer(t *testing.T) {
...
+				reqCtx, cancel := leiosFetchRequestContext(
+					parent,
+					start.Add(tc.deadline),
+				)
+				defer cancel()
+
+				// Advance the fake clock past the requested deadline without
+				// ever cancelling base. A correct context.WithCancel(parent)
+				// result has no expiry of its own and must still be open.
+				time.Sleep(tc.deadline + time.Second)
```
[the remark this answers is not in the document]
Comment (chrisguiney) [reply ouroboros/leios_certified_recovery_test.go:838]: Not needed here. `time.Sleep(tc.deadline + time.Second)` already sleeps past the point where an independent deadline timer (under the reverted/mutated `leiosFetchRequestContext`) would fire, and `testing/synctest` only advances its fake clock past an earlier event once every goroutine woken by that event has itself gone durably blocked or exited. Verified two ways:

1. Reverted `leiosFetchRequestContext` in place to the unconditional `context.WithDeadline(parent, deadline)` this test guards against, with no other change: `go test -race -run TestLeiosFetchRequestContextDoesNotArmIndependentTimerAtEqualOrEarlierParentDeadline -count=10 ./ouroboros/...` fails 10/10, `reqCtx.Err()` already `context deadline exceeded` at the check, no `synctest.Wait()` needed.
2. A standalone `synctest.Test` probe with one `context.WithDeadline` timer firing before a longer `time.Sleep` in the same bubble observes the deadline error at the check across 20/20 runs with no `Wait()`.

Left as-is.

**Review by cryptodj413 — APPROVED** [review]

**Review by wolf31o2 — APPROVED** [review]
